components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt

```diff
@@ -0,0 +1,137 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.storages
+
+import mozilla.components.support.base.log.logger.Logger
+
+import android.support.annotation.VisibleForTesting
+import org.json.JSONArray
+import org.json.JSONObject
+
+/**
+ * This singleton handles the in-memory storage logic for keeping track of
+ * the active experiments. It is meant to be used by through the methods in
+ * the Glean class (General API). No validation on the stored data is
+ * performed at this point: validation must be performed by the General API.
+ */
+internal object ExperimentsStorageEngine : StorageEngine {
+    private val logger = Logger("glean/ExperimentsStorageEngine")
+
+    // Maximum length of the experiment and branch names, in UTF8 byte
+    // sequence length
+    private const val MAX_ID_LENGTH = 30
+
+    data class RecordedExperimentData(
+        val branch: String,
+        val extra: Map<String, String>? = null
+    )
+
+    private val experiments: MutableMap<String, RecordedExperimentData> = mutableMapOf()
+
+    /**
+     * Record an experiment as active.
+     *
+     * @param experimentId the id of the experiment
+     * @param branch the active branch of the experiment
+     * @param extra an optional, user defined String to String map used to provide richer experiment
+     *              context if needed
+     */
+    fun setExperimentActive(
+        experimentId: String,
+        branch: String,
+        extra: Map<String, String>? = null
+    ) {
+        val truncatedExperimentId = experimentId.let {
+            if (it.length > MAX_ID_LENGTH) {
+                logger.warn("experimentId ${it.length} > $MAX_ID_LENGTH")
+                return@let it.substring(0, MAX_ID_LENGTH)
+            }
+            it
+        }
+
+        val truncatedBranch = branch.let {
+            if (it.length > MAX_ID_LENGTH) {
+                logger.warn("branch ${it.length} > $MAX_ID_LENGTH")
+                return@let it.substring(0, MAX_ID_LENGTH)
+            }
+            it
+        }
+
+        val experimentData = RecordedExperimentData(truncatedBranch, extra)
+        experiments.put(truncatedExperimentId, experimentData)
+    }
+
+    /**
+     * Set an active experiment as inactive.
+     *
+     * Warns to the logger (but does not raise) if the experiment is not already active.
+     *
+     * @param experimentId the id of the experiment to set as inactive
+     */
+    fun setExperimentInactive(experimentId: String) {
```

Comment:
We should have test coverage for this and also check on edge cases, such as removing a non-existing experiment.